// nexmo/lib/Account.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _index = require("./index");

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Account = function () {
  /**
   * @param {Credentials} credentials
   *    credentials to be used when interacting with the API.
   * @param {Object} options
   *    Addition Account options.
   */
  function Account(credentials) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Account);

    this.creds = credentials;
    this.options = options;

    // Used to facilitate testing of the call to the underlying object
    this._nexmo = this.options.nexmoOverride || _index2.default;

    this._nexmo.initialize(this.creds.apiKey, this.creds.apiSecret, this.options);
  }

  /**
   * TODO: document
   */


  _createClass(Account, [{
    key: "checkBalance",
    value: function checkBalance(callback) {
      return this.options.rest.get("/account/get-balance", callback);
    }
  }, {
    key: "updatePassword",
    value: function updatePassword(newSecret, callback) {
      return this.options.rest.postUseQueryString("/account/settings", { newSecret: newSecret }, callback);
    }
  }, {
    key: "updateSMSCallback",
    value: function updateSMSCallback(moCallBackUrl, callback) {
      return this.options.rest.postUseQueryString("/account/settings", { moCallBackUrl: moCallBackUrl }, callback);
    }
  }, {
    key: "updateDeliveryReceiptCallback",
    value: function updateDeliveryReceiptCallback(drCallBackUrl, callback) {
      return this.options.rest.postUseQueryString("/account/settings", { drCallBackUrl: drCallBackUrl }, callback);
    }
  }, {
    key: "topUp",
    value: function topUp(trx, callback) {
      return this.options.rest.postUseQueryString("/account/top-up", { trx: trx }, callback);
    }
  }, {
    key: "listSecrets",
    value: function listSecrets(apiKey, callback) {
      return this.options.api.get("/accounts/" + apiKey + "/secrets", {}, callback, false, true);
    }
  }, {
    key: "getSecret",
    value: function getSecret(apiKey, id, callback) {
      return this.options.api.get("/accounts/" + apiKey + "/secrets/" + id, {}, callback, false, true);
    }
  }, {
    key: "createSecret",
    value: function createSecret(apiKey, secret, callback) {
      return this.options.api.postJson("/accounts/" + apiKey + "/secrets/", { secret: secret }, callback, false, true);
    }
  }, {
    key: "deleteSecret",
    value: function deleteSecret(apiKey, id, callback) {
      return this.options.api.delete("/accounts/" + apiKey + "/secrets/" + id, callback, false, true);
    }
  }]);

  return Account;
}();

exports.default = Account;
module.exports = exports["default"];
