// nexmo/lib/Conversion.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _index = require("./index");

var _index2 = _interopRequireDefault(_index);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Conversion = function () {
  function Conversion(credentials, options) {
    _classCallCheck(this, Conversion);

    this.creds = credentials;
    this.options = options;

    // Used to facilitate testing of the call to the underlying object
    this._nexmo = this.options.nexmoOverride || _index2.default;

    this._nexmo.initialize(this.creds.apiKey, this.creds.apiSecret, this.options);
  }

  _createClass(Conversion, [{
    key: "voice",
    value: function voice(message_id, delivered, timestamp, callback) {
      return this.submit("voice", message_id, delivered, timestamp, callback);
    }
  }, {
    key: "sms",
    value: function sms(message_id, delivered, timestamp, callback) {
      return this.submit("sms", message_id, delivered, timestamp, callback);
    }
  }, {
    key: "submit",
    value: function submit(type, message_id, delivered, timestamp, callback) {
      return this.options.api.postUseQueryString("/conversions/" + type, { "message-id": message_id, delivered: delivered, timestamp: timestamp }, this.options.api._addLimitedAccessMessageToErrors(callback, 402));
    }
  }]);

  return Conversion;
}();

exports.default = Conversion;
module.exports = exports["default"];
